src/main/java/bio/terra/service/profile/ProfileService.java

```diff
@@ -212,4 +225,47 @@ public void verifyAccount(BillingProfileRequestModel request, AuthenticatedUserR
         }
     }
 
+    // -- profile upgrade --
+
+    // Billing profiles prior to the introduction of this service did not have sam resources behind them.
+    // This code generates sam resources for any billing profile it does not have one. It sets the
+    // stewards group as the owner of the billing profile.
+    //
+    // We assume this is being run as a Steward and that the Steward has access to any billing
+    public UpgradeResponseModel upgradeProfileResources(UpgradeModel request, AuthenticatedUserRequest user) {
+        Instant startTime = Instant.now();
+        List<BillingProfileModel> profiles = profileDao.getOldBillingProfiles();
+        List<UUID> resources = iamService.listAuthorizedResources(user, IamResourceType.SPEND_PROFILE);
+        Set<String> profileResourceSet;
+        if (resources == null) {
+            profileResourceSet = Collections.EMPTY_SET;
+        } else {
+            profileResourceSet = resources.stream().map(r -> toString()).collect(Collectors.toSet());
+        }
+
+        for (BillingProfileModel profile : profiles) {
```

Comment:
I'm curious to see if this causes an overload in SAM...I know we had to put caching in...though presumably that was for stuff that was happening at a much much higher frequency

Reply:
Never mind, I see it's got a retry under the covers that should help

Reply:
Yes, but for all of our deployments we have a very small number of profiles (< 10) so i think this will be just fine.